---
title: "Homework 2"
author: Philipp Ross
date: "June 11 2015"
output: html_document
---

#### How many protein sequences are there fort he Ebola virus sequencing paper?

Running the following commands can tell us that:
``` {r question1, engine='bash', eval = FALSE}
$ esearch -db protein -query PRJNA257197 | efetch -format fasta > ebola_aa.fasta
$ cat ebola_aa.fasta | grep ">" | wc -l
2240
```

And from this we see that we have **2240** sequences.

#### Unpack at least three SRR runs and see what's inside.

The following data is from Kaneko et al. ([DOI](10.1371/journal.ppat.1004905)) where they looked to characterize the regulatory behavior of a transcription factor found within the unicellular eukaryote, Plasmodium berghei - an agent of malaria that infects mice. The transcription factor is consesrved across species, specifically it is also found within the most deadly of the human infectious malaria, Plasmodium falciparum. 

The data pulled represents an RNA-Seq experiment (SRR1422090), and two biological replicates of ChIP-Seq experiments. (SRR1422094, SRR1422095 and SRR1422096, SRR1422097 - IP and INPUT replicates respectively). I looked at the first 10000 spots using the following commands:

```{r, question2, engine='bash', eval=FALSE}
$ echo -e "SRR1422090\nSRR1422094\nSRR1422095\nSRR1422096\nSRR1422097\n" > sra.ids
$ for run in $(cat sra.ids); do fastq-dump -X 10000 --split-files $run; done
```

I apparently didn't read the run details closely enough because then I would've known that for every experiment they performed single-end sequencing so the *--split-files* flag actually wasn't necessary. I could have also output the result of each dump command to stdout by including the *-Z* flag. The *-X* flag tells fastq-dump that I only want to look at the first X number of spots which are spots on the sequencing flow cell itself. Each spot is a physical location on that flow cell where the sequencer should detect some sort of signal in order to distinguish base calls at different time points in order to form a consensus read for each spot. You can figure this out by multiplying the number of spots designated within the project description run metadata by the read length. Then you should get somewhere around the number of bases that were sequenced which is also listed on the project page next to the number of spots.

After initial inspection of the files and reading through the metadata I can see that the files look a little strange. The transcriptome data is actually from an AB 5500 Genetic Analyzer rather than an Illumina instrument which is what I'm used to:

```{r, question2-1, engine='bash', eval=FALSE}
$ head SRR1422090_1.fastq
@SRR1422090.1 1 length=75
T30.........................................................................
+SRR1422090.1 1 length=75
!..!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@SRR1422090.2 2 length=75
T13.........................................................................
+SRR1422090.2 2 length=75
!..!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
@SRR1422090.3 3 length=75
T12.........................................................................

```

On the other hand the reads generated by the Illumina Genome Analyzer II look normal at first glance:

```{r, question2-2, engine='bash', eval=FALSE}
$ head SRR1422094_1.fastq
@SRR1422094.1 HWI-EAS107_8233_FC30D5P_1_6_1588_1306 length=35
AATCACACTACATAAATAGTGTNNGTANNANTNNN
+SRR1422094.1 HWI-EAS107_8233_FC30D5P_1_6_1588_1306 length=35
IIII>I?II;I=7AI>I1I'I:""+08""0"9"""
@SRR1422094.2 HWI-EAS107_8233_FC30D5P_1_6_1078_354 length=35
ATCCTTAACCCTAACCCAACCCNNACTNNANCNNN
+SRR1422094.2 HWI-EAS107_8233_FC30D5P_1_6_1078_354 length=35
III1I==@-05;/76FH1.1-2""-*&""+"3"""
@SRR1422094.3 HWI-EAS107_8233_FC30D5P_1_6_535_1928 length=35
GTATTTAATAGACTTAACACATNNTTTNNANTNNN
```

##### Several minutes later...

The FASTQ file retrieved from experiments run on a SOLiD instrument are output in what's called color space. The first couple of reads, for whatever reason, are of really low quality which is what we see with the "!" quality score as well as the "." base call. Since the color space file is encoded with Sanger (phred+33) quality scores, this means that the call at this position received a quality score of 0. The "." character apparently stands for an N base call. You can also see that if you add the *-B* flag to the fastq-dump command to force the output to translate the file into base space rather than color space.

Now I can see if I look at the end of the file that the reads are of high enough quality to at least distinguish which base falls at which position within each read:
  
```{r, question2-3, engine='bash', eval=FALSE}
$ fastq-dump -X 10000 -B SRR1422090
$ tail SRR1422090.fastq
+SRR1422090.9998 9998 length=75
DEBBB<<<<<<88(88;((((((.((((((((((((((((..(((((....((((((((((((((((((((((((
@SRR1422090.9999 9999 length=75
TTATGACTTTATATGATTAGAAGGAAAAGAAAAAAGACATGTTACACTTCCTGGTATGTCTTATTACTCATAACC
+SRR1422090.9999 9999 length=75
GHFFFFHHHHHHGGGGHGGGGHGGGGHGGGGHFFFEHGGGEE<<<88838(((((((((((((((((((((((((
@SRR1422090.10000 10000 length=75
AGAATATAAAAAAAGAAAGCAAGGTATAATTGCAGGATATTATGGATCATGGTTAGGCCAAGGAGTTAGCAGGAA
+SRR1422090.10000 10000 length=75
HGBBBA@>>>>;;;;.EDDD>>>>>....38;3333.....33..(......((('....(....((((((''''
```